### PR TITLE
replace SelectGossipOption with SelectOptionByIndex

### LINKED ISSUE #127: [ISSUE] - Gossip frame autoselects are bugged due to API change in 10.0.7
### Description

It looks like a change in the API in 10.0.7 is leading to a lua error on autoselects when player is presented with a gossip frame.

See https://wowpedia.fandom.com/wiki/Patch_10.0.7/API_changes

| 10.0.5 (47825) → 10.0.7 (47983) Feb 5 2023
| ADDED
| ...
| C_GossipInfo.SelectOptionByIndex
```
GossipOptionUIInfo
  # 1: gossipOptionID, Nilable: false -> true
```
If I don't let it autoaccept by holding Control and manually selecting the gossip option, it does not produce a LUA error.

### APR Version

APR 2.2.0

### APR Route

Speedrun

### World of Warcraft Flavor

Retail (Default)

### Tested with only APR

- [ ] Yes
- [X] No

### Lua Error

```true
5x FrameXML/GossipFrameShared.lua:268: attempt to call method 'SelectOptionByIndex' (a nil value)
[string "@FrameXML/GossipFrameShared.lua"]:268: in function `SelectGossipOption'
[string "@APR-Core/QuestHandler.lua"]:3336: in function <APR-Core/QuestHandler.lua:3058>

Locals:
index = 1
(*temporary) = nil
(*temporary) = GossipFrame {
 0 = <userdata>
 SetPortraitShown = <function> defined @SharedXML/PortraitFrame.lua:102
 GetTitleText = <function> defined @SharedXML/PortraitFrame.lua:28
 gossipOptions = <table> {
 }
 SetPortraitTexCoord = <function> defined @SharedXML/PortraitFrame.lua:98
 tutorialButtons = <table> {
 }
 SetPortraitTextureRaw = <function> defined @SharedXML/PortraitFrame.lua:84
 OnShow = <function> defined @FrameXML/GossipFrameShared.lua:149
 hasActiveQuests = true
 buttons = <table> {
 }
 AvailableQuestButtonInit = <function> defined @FrameXML/GossipFrameShared.lua:102
 SetTitleMaxLinesAndHeight = <function> defined @SharedXML/PortraitFrame.lua:49
 SetTitleColor = <function> defined @SharedXML/PortraitFrame.lua:32
 SetPoint = <function> defined =[C]:-1
 backdrop = Frame {
 }
 Update = <function> defined @FrameXML/GossipFrameShared.lua:208
 FriendshipStatusBar = StatusBar {
 }
 GreetingPanel = Frame {
 }
 SetTitle = <function> defined @SharedXML/PortraitFrame.lua:36
 OnEvent = <function> defined @FrameXML/GossipFrame.lua:81
 SetTitleOffsets = <function> defined @SharedXML/PortraitFrame.lua:54
 Inset = GossipFrameInset {
 }
 CloseButton = GossipFrameCloseButton {
 }
 TopTileStreaks = Texture {
 }
 SetGossipTitle = <function> defined @FrameXML/GossipFrameShared.lua:261
 SetBorder = <function> defined @SharedXML/PortraitFrame.lua:67
 layoutType = "PortraitFrameTemplate"
 TitleContainer = Frame {
 }
 PortraitContainer = Frame {
 }
 SetFrameLevelsFromBaseLevel = <function> defined @SharedXML/PortraitFrame.lua:120
 SetPortraitToBag = <function> defined @SharedXML/PortraitFrame.lua:80
 SetPortraitAtlasRaw = <function> defined @SharedXML/PortraitFrame.lua:88
 SetPortraitToUnit = <function> defined @SharedXML/PortraitFrame.lua:76
 Background = Texture {
 }
 Bg = GossipFrameBg {
 }
 SetTitleFormatted = <function> defined @SharedXML/PortraitFrame.lua:40
 GetPortrait = <function> defined @SharedXML/PortraitFrame.lua:63
 UpdateScrollBox = <function> defined @FrameXML/GossipFrameShared.lua:109
 NineSlice = Frame {
 }
 OnLoad = <function> defined @FrameXML/GossipFrame.lua:56
 OnHide = <function> defined @FrameXML/GossipFrameShared.lua:156
 SetPortraitToAsset = <function> defined @SharedXML/PortraitFrame.lua:72
 SetPortraitTextureSizeAndOffset = <function> defined @SharedXML/PortraitFrame.lua:106
 HandleShow = <function> defined =[C]:-1
 SortOrder = <function> defined @FrameXML/GossipFrame.lua:97
 SelectGossipOption = <function> defined @FrameXML/GossipFrameShared.lua:191
 HandleHide = <function> defined @FrameXML/GossipFrameShared.lua:186
 GetTutorialButtons = <function> defined @FrameXML/GossipFrame.lua:93
 SetGossipTutorialMode = <function> defined @FrameXML/GossipFrame.lua:87
 GetBackgroundTexture = <function> defined @FrameXML/GossipFrame.lua:70
 SetPortraitToClassIcon = <function> defined @SharedXML/PortraitFrame.lua:92
}
(*temporary) = 1
(*temporary) = "attempt to call method 'SelectOptionByIndex' (a nil value)"
```

### Reproduction Steps

1. Speak to an NPC that has a gossip option or a quest pickup/turnin where the frame uses a gossip frame
2. Receive LUA error

### Last Good Version

APR 2.1.2

### Screenshots

image

### Any additional information?

_No response_

### PR DESCRIPTION
**Pull request by oliver-cool-beans**
close #127 
Updated to C_GossipInfo.SelectOptionByIndex
and uncomment replacement codeblock

### CODE REVIEW

File: APR-Core/QuestHandler.lua [hunk 1]
```diff
@@ -3283,68 +3283,68 @@ APR_QH_EventFrame:SetScript("OnEvent", function(self, event, ...)
...
-					-- 	for i, v in pairs(info) do
-					-- 		if(v.orderIndex+1 == steps["Gossip"]) then
-					-- 			C_GossipInfo.SelectOption(v.gossipOptionID)
-					-- 		end
-					-- 	end
-					-- else
-					-- 	SelectGossipOption(steps["Gossip"])
-					-- end
+					-- C_GossipInfo.SelectOptionByIndex(steps["Gossip"])
+					local info = C_GossipInfo.GetOptions()
```
Comment (Neogeekmo) [inline APR-Core/QuestHandler.lua:3338]: I was right to make this loop, I had smelled the change coming xD

**Review by Neogeekmo — APPROVED** [review]

File: APR-Core/QuestHandler.lua [hunk 2]
```diff
@@ -3283,68 +3283,68 @@ APR_QH_EventFrame:SetScript("OnEvent", function(self, event, ...)
 			-- GOSSIP HARDCODED
 			if (steps and steps["Gossip"]) then
 				if (steps["Gossip"] == 27373 or steps["Gossip"] == 34398) then
-					SelectGossipOption(1)
+					C_GossipInfo.SelectOptionByIndex(1)
```
Comment (Neogeekmo) [inline APR-Core/QuestHandler.lua:3286]: I don't know if it's this value or -1 of the value (we keep it and we'll do a fix if ever we get a bug)

File: APR-Core/QuestHandler.lua [hunk 3]
```diff
@@ -3283,68 +3283,68 @@ APR_QH_EventFrame:SetScript("OnEvent", function(self, event, ...)
...
+					C_GossipInfo.SelectOptionByIndex(2)
 					APR1[APR.Realm][APR.Name][APR.ActiveMap] = APR1[APR.Realm][APR.Name][APR.ActiveMap] + 1
 					APR.BookingList["UpdateQuest"] = 1
 					APR.BookingList["PrintQStep"] = 1
 				end
 				if (steps["Gossip"] == 28202) then
 					APRGOSSIPCOUNT = APRGOSSIPCOUNT + 1
 					if (APRGOSSIPCOUNT == 1) then
-						SelectGossipOption(1)
+						C_GossipInfo.SelectOptionByIndex(1)
```
Comment (Neogeekmo) [inline APR-Core/QuestHandler.lua:3300]: I don't know if it's this value or -1 of the value (we keep it and we'll do a fix if ever we get a bug)

File: APR-Core/QuestHandler.lua [hunk 4]
```diff
@@ -3283,68 +3283,68 @@ APR_QH_EventFrame:SetScript("OnEvent", function(self, event, ...)
...
 					elseif (APRGOSSIPCOUNT == 2) then
 						if (APR.Race == "Gnome") then
-							SelectGossipOption(1)
+							C_GossipInfo.SelectOptionByIndex(1)
```
Comment (Neogeekmo) [inline APR-Core/QuestHandler.lua:3303]: I don't know if it's this value or -1 of the value (we keep it and we'll do a fix if ever we get a bug)

File: APR-Core/QuestHandler.lua [hunk 5]
```diff
@@ -3283,68 +3283,68 @@ APR_QH_EventFrame:SetScript("OnEvent", function(self, event, ...)
...
 						elseif (APR.Race == "Human" or APR.Race == "Dwarf") then
-							SelectGossipOption(2)
+							C_GossipInfo.SelectOptionByIndex(2)
```
Comment (Neogeekmo) [inline APR-Core/QuestHandler.lua:3305]: I don't know if it's this value or -1 of the value (we keep it and we'll do a fix if ever we get a bug)

File: APR-Core/QuestHandler.lua [hunk 6]
```diff
@@ -3283,68 +3283,68 @@ APR_QH_EventFrame:SetScript("OnEvent", function(self, event, ...)
...
 						elseif (APR.Race == "NightElf") then
-							SelectGossipOption(3)
+							C_GossipInfo.SelectOptionByIndex(3)
```
Comment (Neogeekmo) [inline APR-Core/QuestHandler.lua:3307]: I don't know if it's this value or -1 of the value (we keep it and we'll do a fix if ever we get a bug)

File: APR-Core/QuestHandler.lua [hunk 7]
```diff
@@ -3283,68 +3283,68 @@ APR_QH_EventFrame:SetScript("OnEvent", function(self, event, ...)
...
 						elseif (APR.Race == "Draenei" or APR.Race == "Worgen") then
-							SelectGossipOption(4)
+							C_GossipInfo.SelectOptionByIndex(4)
```
Comment (Neogeekmo) [inline APR-Core/QuestHandler.lua:3309]: I don't know if it's this value or -1 of the value (we keep it and we'll do a fix if ever we get a bug)

File: APR-Core/QuestHandler.lua [hunk 8]
```diff
@@ -3283,68 +3283,68 @@ APR_QH_EventFrame:SetScript("OnEvent", function(self, event, ...)
...
 						end
 					elseif (APRGOSSIPCOUNT == 3) then
 						if (APR.Race == "Gnome") then
-							SelectGossipOption(3)
+							C_GossipInfo.SelectOptionByIndex(3)
```
Comment (Neogeekmo) [inline APR-Core/QuestHandler.lua:3313]: I don't know if it's this value or -1 of the value (we keep it and we'll do a fix if ever we get a bug)

File: APR-Core/QuestHandler.lua [hunk 9]
```diff
@@ -3283,68 +3283,68 @@ APR_QH_EventFrame:SetScript("OnEvent", function(self, event, ...)
...
 						elseif (APR.Race == "Human" or APR.Race == "Dwarf") then
-							SelectGossipOption(4)
+							C_GossipInfo.SelectOptionByIndex(4)
```
Comment (Neogeekmo) [inline APR-Core/QuestHandler.lua:3315]: I don't know if it's this value or -1 of the value (we keep it and we'll do a fix if ever we get a bug)

File: APR-Core/QuestHandler.lua [hunk 10]
```diff
@@ -3283,68 +3283,68 @@ APR_QH_EventFrame:SetScript("OnEvent", function(self, event, ...)
...
 						elseif (APR.Race == "NightElf") then
-							SelectGossipOption(2)
+							C_GossipInfo.SelectOptionByIndex(2)
 						elseif (APR.Race == "Draenei" or APR.Race == "Worgen") then
-							SelectGossipOption(1)
+							C_GossipInfo.SelectOptionByIndex(1)
```
Comment (Neogeekmo) [inline APR-Core/QuestHandler.lua:3319]: I don't know if it's this value or -1 of the value (we keep it and we'll do a fix if ever we get a bug)

File: APR-Core/QuestHandler.lua [hunk 11]
```diff
@@ -3283,68 +3283,68 @@ APR_QH_EventFrame:SetScript("OnEvent", function(self, event, ...)
...
 						end
 					elseif (APRGOSSIPCOUNT == 4) then
 						if (APR.Race == "Gnome") then
-							SelectGossipOption(4)
+							C_GossipInfo.SelectOptionByIndex(4)
```
Comment (Neogeekmo) [inline APR-Core/QuestHandler.lua:3323]: I don't know if it's this value or -1 of the value (we keep it and we'll do a fix if ever we get a bug)

File: APR-Core/QuestHandler.lua [hunk 12]
```diff
@@ -3283,68 +3283,68 @@ APR_QH_EventFrame:SetScript("OnEvent", function(self, event, ...)
...
 						elseif (APR.Race == "Human" or APR.Race == "Dwarf") then
-							SelectGossipOption(2)
+							C_GossipInfo.SelectOptionByIndex(2)
```
Comment (Neogeekmo) [inline APR-Core/QuestHandler.lua:3325]: I don't know if it's this value or -1 of the value (we keep it and we'll do a fix if ever we get a bug)

File: APR-Core/QuestHandler.lua [hunk 13]
```diff
@@ -3283,68 +3283,68 @@ APR_QH_EventFrame:SetScript("OnEvent", function(self, event, ...)
...
 						elseif (APR.Race == "NightElf") then
-							SelectGossipOption(1)
+							C_GossipInfo.SelectOptionByIndex(1)
```
Comment (Neogeekmo) [inline APR-Core/QuestHandler.lua:3327]: I don't know if it's this value or -1 of the value (we keep it and we'll do a fix if ever we get a bug)

**Review by NightofStarrs — APPROVED** [review]
As mentioned, if this causes issues because of the index value we can adjust as necessary.

**Pull request merged**